File: backend/domain/entity/case.go
package entity

import "time"

// DecisionCase is the aggregate root of a MAGI decision run.
type DecisionCase struct {
	ID           string
	UserID       int64
	Question     string
	Context      string
	Constraints  []Constraint
	ParentCaseID string // lineage link: this case was forked from ParentCaseID
	Status       CaseStatus
	CurrentPhase CasePhase
	// Pinned/Archived are user-level list management flags (P2 D11): pin keeps a
	// case at the top of the sidebar, archive hides it from active lists.
	Pinned   bool
	Archived bool
	// ExecutionAttempt is set by the durable worker for the current attempt.
	// It is intentionally runtime-only: the durable job owns the source of truth.
	ExecutionAttempt int
	// PausedFromStatus records the case status before a task-level pause, so
	// wake can continue the FSM from where it stopped.
	PausedFromStatus CaseStatus
	MaxDebateRounds  int
	Deadline         *time.Time
	CreatedAt        time.Time
	UpdatedAt        time.Time
}

type CaseStatus string

const (
	CaseStatusDraft            CaseStatus = "DRAFT"
	CaseStatusNormalizing      CaseStatus = "NORMALIZING"
	CaseStatusContextBuilding  CaseStatus = "CONTEXT_BUILDING"
	CaseStatusRetrievingMemory CaseStatus = "RETRIEVING_MEMORY"
	CaseStatusInvestigating    CaseStatus = "INVESTIGATING"
	CaseStatusEvidenceGating   CaseStatus = "EVIDENCE_GATING"
	CaseStatusCollectingVotes  CaseStatus = "COLLECTING_VOTES"
	CaseStatusConsensusCheck   CaseStatus = "CONSENSUS_CHECK"
	CaseStatusResolving        CaseStatus = "RESOLVING"
	CaseStatusGeneratingReport CaseStatus = "GENERATING_REPORT"
	CaseStatusSavingMemory     CaseStatus = "SAVING_MEMORY"
	CaseStatusEvaluating       CaseStatus = "EVALUATING"
	CaseStatusDebating         CaseStatus = "DEBATING"
	CaseStatusReflecting       CaseStatus = "REFLECTING"
	CaseStatusRevoting         CaseStatus = "REVOTING"
	CaseStatusResolved         CaseStatus = "RESOLVED"
	CaseStatusMemoryIndexed    CaseStatus = "MEMORY_INDEXED"
	CaseStatusFailed           CaseStatus = "FAILED"
	CaseStatusCancelled        CaseStatus = "CANCELLED"
	CaseStatusTimedOut         CaseStatus = "TIMED_OUT"
	CaseStatusInsufficientEv   CaseStatus = "INSUFFICIENT_EVIDENCE"
	CaseStatusDeadlocked       CaseStatus = "DEADLOCKED"
	CaseStatusPaused           CaseStatus = "PAUSED"
)

type CasePhase string

// Constraint is a user-supplied decision constraint.
type Constraint struct {
	Key   string `json:"key"`
	Value string `json:"value"`
	Hard  bool   `json:"hard"`
}

File: frontend/src/types/case.ts
export type CaseStatus =
  | 'DRAFT'
  | 'NORMALIZING'
  | 'CONTEXT_BUILDING'
  | 'RETRIEVING_MEMORY'
  | 'INVESTIGATING'
  | 'EVIDENCE_GATING'
  | 'COLLECTING_VOTES'
  | 'CONSENSUS_CHECK'
  | 'DEBATING'
  | 'REFLECTING'
  | 'REVOTING'
  | 'RESOLVING'
  | 'GENERATING_REPORT'
  | 'SAVING_MEMORY'
  | 'EVALUATING'
  | 'RESOLVED'
  | 'DEADLOCKED'
  | 'FAILED'
  | 'CANCELLED'
  | 'TIMED_OUT'
  | 'INSUFFICIENT_EVIDENCE'
  | 'MEMORY_INDEXED';

export const CASE_STATUS_LABELS: Record<CaseStatus, string> = {
  DRAFT: 'Draft',
  NORMALIZING: 'Normalizing',
  CONTEXT_BUILDING: 'Building Context',
  RETRIEVING_MEMORY: 'Retrieving Memory',
  INVESTIGATING: 'Investigating',
  EVIDENCE_GATING: 'Evidence Gate',
  COLLECTING_VOTES: 'Collecting Votes',
  CONSENSUS_CHECK: 'Consensus Check',
  DEBATING: 'Debating',
  REFLECTING: 'Reflecting',
  REVOTING: 'Re-voting',
  RESOLVING: 'Resolving',
  GENERATING_REPORT: 'Generating Report',
  SAVING_MEMORY: 'Saving Memory',
  EVALUATING: 'Evaluating',
  RESOLVED: 'Resolved',
  DEADLOCKED: 'Deadlocked',
  FAILED: 'Failed',
  CANCELLED: 'Cancelled',
  TIMED_OUT: 'Timed Out',
  INSUFFICIENT_EVIDENCE: 'Insufficient Evidence',
  MEMORY_INDEXED: 'Memory Indexed',
};

export interface Case {
  id: string;
  question: string;
  background: string;
  constraints: Constraint[];
  status: CaseStatus;
  round: number;
  consensus: ConsensusState | null;
  confidence: number;
  finalDecision: string;
  createdAt: string;
  updatedAt: string;
}

export interface CaseSummary {
  id: string;
  question: string;
  status: CaseStatus;
  round: number;
  createdAt: string;
  pinned: boolean;
}

export interface Constraint {
  label: string;
  value: string;
}

export interface ConsensusState {
  approve: number;
  reject: number;
  abstain: number;
  majority: 'Approve' | 'Reject' | 'Tie';
  needReflection: boolean;
}
